vendor/src/github.com/microsoft/hcsshim/createprocess.go

```diff
@@ -12,30 +14,71 @@ import (
 // processParameters is use to both the input of CreateProcessInComputeSystem
 // and to convert the parameters to JSON for passing onto the HCS
 type CreateProcessParams struct {
-	ApplicationName                   string
-	CommandLine                       string
-	WorkingDirectory                  string
-	StdInPipe, StdOutPipe, StdErrPipe string
-	Environment                       map[string]string
-	EmulateConsole                    bool
-	ConsoleSize                       [2]int
+	ApplicationName  string
+	CommandLine      string
+	WorkingDirectory string
+	Environment      map[string]string
+	EmulateConsole   bool
+	ConsoleSize      [2]int
+}
+
+// pipe struct used for the stdin/stdout/stderr pipes
+type pipe struct {
+	handle syscall.Handle
+}
+
+func makePipe(h syscall.Handle) *pipe {
+	p := &pipe{h}
+	runtime.SetFinalizer(p, (*pipe).closeHandle)
+	return p
+}
+
+func (p *pipe) closeHandle() {
+	if p.handle != 0 {
+		syscall.CloseHandle(p.handle)
+		p.handle = 0
+	}
+}
+
+func (p *pipe) Close() error {
+	p.closeHandle()
+	runtime.SetFinalizer(p, nil)
+	return nil
+}
+
+func (p *pipe) Read(b []byte) (int, error) {
+	// syscall.Read returns 0, nil on ERROR_BROKEN_PIPE, but for
+	// our purposes this should indicate EOF. This may be a go bug.
```

Comment:
Any chance of making this work with file descriptors? You could just use http://golang.org/pkg/os/#NewFile and then avoid having to deal with these edges cases.

I could be way off: this would require that `syscall.Handle` is a `uintptr`. A cursory glance should that this might work.